#!/bin/bash

#PBS -l nodes=1:ppn=8,walltime=40:00:00
#PBS -N gatk_geno_indiv
#PBS -m abe
#PBS -e localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.e${PBS_JOBID}.${PBS_ARRAYID}
#PBS -o localhost:${PBS_O_WORKDIR}/${PBS_JOBNAME}.o${PBS_JOBID}.${PBS_ARRAYID}

# ------------------------------------------------------------------------------
# Variables
# ------------------------------------------------------------------------------

# Call from NGS-map

working_dir=$PBS_O_WORKDIR

module load jdk/1.7.0_60

GATK=~/exome_macaque/bin/GATK
GENOME_FA=genomes/papAnu2/papAnu2.fa
GENOME_NAME=baboon

module load vcftools
module load tabix

# ------------------------------------------------------------------------------
# Run program
# ------------------------------------------------------------------------------

cd $working_dir

if [ "$PBS_ARRAYID" -eq 21 ]; then
    CHROM=X
else
    CHROM=$PBS_ARRAYID
fi

BAMS=(`ls results/*.PE.bwa.${GENOME_NAME}.passed.realn.bam`)

# Make output directory
mkdir ${GENOME_NAME}_snps

count=0
for b in ${BAMS[*]}; do

    count=`expr $count + 1`

    java -jar ${GATK}/GenomeAnalysisTK.jar \
        -T UnifiedGenotyper \
        -R ${GENOME_FA} \
        -I ${b} \
        -stand_call_conf 50.0 \
        -stand_emit_conf 10.0 \
        -out_mode EMIT_ALL_CONFIDENT_SITES \
        -o ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.tmp${count}.vcf \
        -nct 4 \
        -nt 8 \
        -L chr${CHROM}
    
    bgzip ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.tmp${count}.vcf
    tabix -p vcf ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.tmp${count}.vcf.gz

done

# Merge this chromosome's VCFs
TMP_VCFS=(`ls ${GENOME_NAME}_snps/chr${CHROM}.raw.snps.indels.tmp*.vcf.gz`)
vcf-merge ${TMP_VCFS[*]} > ${GENOME_NAME}_snps/chr${CHROM}.INDIV.raw.snps.indels.vcf

exit;
